### ISSUE TITLE
ADM5 Pro Filtration Fans are backwards and servo missing

### ISSUE THREAD

**Issue by johncotten** [xblax/flashforge_adm5_klipper_mod#68]
The internal and external fans are flipped. Pin PB8 should be External, PB6 should be Internal. There is also a servo that flips the door when you engage External Filtration(fan). The default behavior for the servo is flipped down to only circulate internally. Both of the Pro's I received actually had the door installed upright which did not allow the servo to move. You can hear the server being activated on Nathan Builds Robots review of the pro when he's checking the filtration system. So internal filtration, which is just the blower to return air back to the chamber, should only be PB6. But External Filtration engages the servo, then both the Internal(blower) and External(4020) fans. Hope this makes sense.

**Comment by consp** [comment]
> Pin PB8 should be External, PB6 should be Internal. 

Maybe that's an issue with the AD5M non-pro config. @xblax do you know if there maybe is a difference between the original two configs? (just a wild guess)

See also #53 and #54 for reference since it's related. Would be good to add some info about this production issue somewhere since it's broken from factory by the looks of it.

**Comment by KaruroChori** [comment]
I think that in order to properly calibrate it, and to fully implement some of the original functionalities, we need the base "temperature" as reference, and maybe the level you get while printing abs while enclosed. This should provide some realistic values to use.

In terms of visualization, the fact it is show as a temperature value is not a real problem except that it could be confusing for the end user. I have no idea if there is a way to change visualization and units.

**Comment by xblax** [comment]
@johncotten
Can you maybe attach the original klipper config from the 5M Pro? That is at `/mnt/orig_root/opt/config/printer.base.cfg`
Would be interesting if the fans are named wrong there as well.

I think we should rename `macros.cfg` to macros_adm5.cfg and add additional macros_adm5_pro.cfg that has the Pro specific macros. The 5M Pro will have to include both files from printer.cfg

**Comment by consp** [comment]
Can now confirm that pins are reversed, weird.

**Comment by consp** [comment]
>  Default behavior is flap is closed for internal only, 0 degrees or at the 3 o'clock position. External, flap opens and moved into the 90 or 12 o'clock position?

Yes, that is what mine came with, 0 == flat (internal circulation), 90 == upright (external). I'd rather have some more confirmation since you already mentioned some from-factory issues. Mine makes nasty noises when you attempt to do 180. Probably n=2 is a bit too few to be sure.

**Comment by consp** [comment]
Internal (was 0, now 90)
![IMG_20240316_103025_MP](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/6802998/27e18da4-3bbd-47d3-9041-554c3d1a9d3c)
External (was 90 now 180):
![IMG_20240316_103042_MP](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/6802998/40e9c3ee-54c0-41a9-8d48-f6189a31f129)
Without door:
![IMG_20240316_105632_MP](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/6802998/627981b6-1bec-4182-8b00-1025d0569ae7)


Long story short: They have some assembly instructions to manage, door was put in the wrong position and as a bonus the screw was to tight for the tollerances of these kind of molds. I greased it up as well and now works as expected.

**Comment by consp** [comment]
> it was probably a fire hazard because it left the servo on and grinding or stuck.

Fortunately the wires will most likely melt and break before fire errupts.

For the servo grinding issue: maybe it's a good idea to set the defaults to 95 or 93 (which was safe for me) and 180. It mostly grinds in "internal" mode at the 90 position. This definately needs testing though by more users (or more printers).

**Comment by Tiwatz** [comment]
> > it was probably a fire hazard because it left the servo on and grinding or stuck.
> 
> Fortunately the wires will most likely melt and break before fire errupts.
> 
> For the servo grinding issue: maybe it's a good idea to set the defaults to 95 or 93 (which was safe for me) and 180. It mostly grinds in "internal" mode at the 90 position. This definately needs testing though by more users (or more printers).

On my servo the internal gears broke. Flashforge sent me a new one but the gear was way too loose on the "door". I bought a metal geared one on Amazon (ordinary sg90) and it was the same thing, way too loose. So I had to take two servos apart and replace the broken piece on the original servo. So I guess the servo will break before any fire will occur.
Btw my door was in the wrong direction as well.

**Comment by consp** [comment]
Added the script and suggestions of splitting up the files. @xblax I don't know what is sensible so created multiple commits with all the different parts.

Changed the default angle to 95deg since on my machine it sometimes starts complaining (and getting hot) at 90.

**Comment by xblax** [comment]
@consp
This looks good, but I will try to avoid the duplicated END_PRINT before adding this. Maybe via a hook that we add to _COMMON_END_PRINT (untested):

```
{printer['gcode_macro _M5PRO_MACRO_HOOKS'].common_end_print|default("")}
```

**Comment by consp** [comment]
I'll have a look, not used to the advanced klipper stuff yet.

**Comment by consp** [comment]
> Contemplating replacing the external fan to pull a lot more air out of there.

yes I'm looking into that as well. Maybe a duct with a 80mm fan. I've got some high static pressure fans laying around which would suite the job. Same for internal, complete replacement.

**Comment by consp** [comment]
> @consp This looks good, but I will try to avoid the duplicated END_PRINT before adding this. Maybe via a hook that we add to _COMMON_END_PRINT (untested):
> 
> ```
> {printer['gcode_macro _M5PRO_MACRO_HOOKS'].common_end_print|default("")}
> ```

I'm actually in favour of not doing the circulation stop at endprint time and leaving it out. Switched to abs printing and I want it to continue running after the print has finished so I added it to the slicer settings instead. Also makes more sense since it's related to the material used.

**Comment by xblax** [comment]
> I'm actually in favour of not doing the circulation stop at endprint time and leaving it out. Switched to abs printing and I want it to continue running after the print has finished so I added it to the slicer settings instead. Also makes more sense since it's related to the material used. see [19abdd4](https://github.com/xblax/flashforge_adm5_klipper_mod/tree/feature/servo_angle)

Yes, makes sense. You could even have a delayed G-Code macro to keep the fans running for a fixed time after printing.

I now included the macros on master.

**Comment by Tiwatz** [comment]
I just noticed in `macros-pro.cfg` line 38 should be `SET_SERVO SERVO=my_servo ANGLE=180`

**Comment by consp** [comment]
> I just noticed in `macros-pro.cfg` line 38 should be `SET_SERVO SERVO=my_servo ANGLE=180`

That seems correct. Since it's quite a nasty bug I've already pushed it to master.

**Comment by MrSolidus89** [comment]
Sorry guys, but is it possible to create the same system for the 5M?  or print the same conveyor and add more fan?

**Comment by consp** [comment]
> Sorry guys, but is it possible to create the same system for the 5M? or print the same conveyor and add servo plus fan?

Haven't seen anyone make one yet, should be possible since the plastic parts and attachment points are the same. My advise will be to make one which does not use a blower fan and has a larger door since it's all a bit tiny.

**Comment by MrSolidus89** [comment]
> > Sorry guys, but is it possible to create the same system for the 5M? or print the same conveyor and add servo plus fan?
> 
> Haven't seen anyone make one yet, should be possible since the plastic parts and attachment points are the same. My advise will be to make one which does not use a blower fan and has a larger door since it's all a bit tiny.

it would be interesting if a 5M PRO owner could replicate the same thing so that we who have the 5M can also take advantage of this filtering system for the closed chamber

**Comment by the-redone** [comment]
Anyone help with this, I've experienced one original servo die with the squealing sound of the motor in the servo dying, replaced with another servo, that also experienced the same after 3 prints, replaced again and that also had the same issue after maybe 15 minutes of just being idle.

Unsure what else to do other than rule it as a main board issue and return the printer. I see people saying the pins are reversed, which ones and does this apply to stock machines?

**Comment by jackpunt** [comment]
Is there a [simple] process to determine if this problem exists on new/current release of A5M Pro?

**Comment by believerd** [comment]
I can confirm those pins are flipped on my 5m pro.

Below is the original config.
```
[fan_generic internal_fan]
pin:PB8

[fan_generic external_fan]
pin:PB6
```

**Comment by jackpunt** [comment]
For the newly arrived, can you explain how/where to find the config file?
(and: can this be resolved by changing the config? or is a mechanical change required?)

**Comment by kevinz2018** [comment]
> Is there a [simple] process to determine if this problem exists on new/current release of A5M Pro?

Would love to know this as well. Just bought a 5M pro a couple of weeks ago. 

Based on my understanding, for the original problem, you need to do two things?
1. flip the pin settings in the original config file.
2. open the back and change the door position 90 degrees?

Is my understanding right?  Or Is it possible to just do either 1 or 2? 

When I manually switch btw internal fan/external fan, I can hear click sound. 

Also in neither mode I can NOT feel any air comes out of the small round external vent on the bottom left.  Is that normal?  or is the air vent so weak that I can't feel it?

Thanks.
